### PR TITLE
feat: implement bucket.uploadDirectory() method

### PR DESCRIPTION
**Pull request by AVaksman**
Towards #540

- [x] Tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)

### CODE REVIEW

File: src/bucket.ts [hunk 1]
```diff
@@ -323,6 +324,16 @@ export interface UploadOptions
   resumable?: boolean;
 }
 
+export interface UploadDirectoryOptions extends CreateWriteStreamOptions {
+  recurse?: boolean;
+}
+
+export interface UploadDirectoryCallback {
+  (err: Error | null, resp?: Array<{[k: string]: string | Error}>): void;
```
Comment (callmehiphop) [inline src/bucket.ts]: Can we move the nested object to its own type?

File: src/bucket.ts [hunk 2]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+    directoryPathString: string,
+    options: UploadDirectoryOptions,
+    callback: UploadDirectoryCallback
+  ): void;
+  uploadDirectory(
+    directoryPathString: string,
+    callback: UploadDirectoryCallback
+  ): void;
+  /**
+   * @typedef {object} UploadDirectoryOptions Configuration options for Bucket#uploadDirectory().
```
Comment (callmehiphop) [inline src/bucket.ts]: Lets wrap this in a `{@link}` tag

File: src/bucket.ts [hunk 3]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+   * @param {boolean} [options.recurse] Whether to recursevely upload all files from
```
Comment (callmehiphop) [inline src/bucket.ts]: I *think* these should be `@property` tags

Comment (callmehiphop) [reply src/bucket.ts]: Also looks like a typo here

File: src/bucket.ts [hunk 4]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+   * @param {string|boolean} [options.validation] Possible values: `"md5"`,
+   *     `"crc32c"`, or `false`. By default, data integrity is validated with an
+   *     MD5 checksum for maximum reliability. CRC32c will provide better
+   *     performance with less reliability. You may also choose to skip
+   * validation completely, however this is **not recommended**.
+   */
+  /**
+   * @typedef {array} UploadDirectoryResponse
+   * @property {object} [response] Response for each of the uploaded files.
+   * @param {string} [filename] Absolute file path.
```
Comment (callmehiphop) [inline src/bucket.ts]: I think these should all be `@property`

File: src/bucket.ts [hunk 5]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+   *     billed for the request.
+   * @param {string|boolean} [options.validation] Possible values: `"md5"`,
+   *     `"crc32c"`, or `false`. By default, data integrity is validated with an
+   *     MD5 checksum for maximum reliability. CRC32c will provide better
+   *     performance with less reliability. You may also choose to skip
+   * validation completely, however this is **not recommended**.
+   */
+  /**
+   * @typedef {array} UploadDirectoryResponse
+   * @property {object} [response] Response for each of the uploaded files.
```
Comment (callmehiphop) [inline src/bucket.ts]: I think since we're describing an array these should be index values.

File: src/bucket.ts [hunk 6]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+   */
+  /**
+   * @callback UploadDirectoryCallback
+   * @param {?Error} err Request error, if any.
+   * @param {array} resp The list of object responses for every uploaded file.
+   */
+
+  /**
+   * Upload a directory to the bucket. This is a convenience method for multiple
+   * {@link Bucket#upload}.
```
Comment (callmehiphop) [inline src/bucket.ts]: I think maybe this should say something like
> This is a convenience method for making multiple `{@link Bucket#upload}` calls.

File: src/bucket.ts [hunk 7]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+   * @returns {Promise<UploadDirectoryResponse>}
+   *
+   * @example
+   * const {Storage} = require('@google-cloud/storage');
+   * const storage = new Storage();
+   * const bucket = storage.bucket('albums');
+   *
+   * //-
+   * // Upload a directory from a local path.
+   * //-
```
Comment (callmehiphop) [inline src/bucket.ts]: The `//-` syntax was for an old version of the docs, we shouldn't be using it anymore. Instead lets use multiple `@example` tags and `<caption>` to add a descriptions to the examples.

File: src/bucket.ts [hunk 8]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+   * bucket.uploadDirectory('/local/path/myfolder', function(err, resp) {
+   *   // Your bucket now contains:
+   *   // - "myfolder/" directory (with the contents of `/local/path/myfolder/')
+   *
+   *   // `file` is an instance of a File object that refers to your new file.
```
Comment (callmehiphop) [inline src/bucket.ts]: Is this leftover from copy/paste? I don't see a `file` variable anywhere

File: src/bucket.ts [hunk 9]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+   * });
+   *
+   *
+   * //-
+   * // You can also have files gzip'd on the fly.
+   * //-
+   * bucket.uploadDirectory(path.join(__dirname, 'index', { gzip: true }, function(err, resp) {
```
Comment (callmehiphop) [inline src/bucket.ts]: We should either `import` path or not use it at all for the example.

File: src/bucket.ts [hunk 10]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+   *   // Your bucket now contains:
+   *   // - "index/" directory (automatically compressed with gzip)
+   *
+   *   // Downloading the file with `file.download` will automatically decode
+   * the
```
Comment (callmehiphop) [inline src/bucket.ts]: looks like this comment needs some formatting love, also I'm a little confused about the context of `file.download`, are we saying any of the files you just uploaded will be automatically decoded?

File: src/bucket.ts [hunk 11]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+   *   // file.
+   * });
+   *
+   *
+   * //-
+   * // If the callback is omitted, we'll return a Promise.
+   * //-
+   * bucket.uploadDirectory('local-directory-path').then(function(data) {
```
Comment (callmehiphop) [inline src/bucket.ts]: Could probably make this a little more modern with

```js
const [response] = await bucket.uploadDirectory('local-directory-path');
```

File: src/bucket.ts [hunk 12]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+   * });
+   */
+
+  uploadDirectory(
+    directoryPathString: string,
+    optionsOrCallback?: UploadDirectoryOptions | UploadDirectoryCallback,
+    callback?: UploadDirectoryCallback
+  ): Promise<UploadDirectoryResponse> | void {
+    // tslint:disable-next-line:no-any
+    if ((global as any)['GCLOUD_SANDBOX_ENV']) {
```
Comment (callmehiphop) [inline src/bucket.ts]: @stephenplusplus maybe we should move this to `process.env.GCLOUD_SANDBOX_ENV`?

Comment (callmehiphop) [reply src/bucket.ts]: Actually do we even need this anymore? I almost think this was from a time where we were smoke testing the JSDoc snippets

File: src/bucket.ts [hunk 13]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+    }
+
+    const options =
+      typeof optionsOrCallback === 'object' ? optionsOrCallback : {};
+    callback =
+      typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
+
+    let stat: fs.Stats | Error;
+    try {
+      stat = fs.lstatSync(directoryPathString);
```
Comment (callmehiphop) [inline src/bucket.ts]: I think we might want to use the async version of this? @bcoe @stephenplusplus thoughts?

File: src/bucket.ts [hunk 14]
```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
...
+    Promise.all(
+      fileList.map(file =>
+        limit(() => {
+          let destination = path.relative(pathDirName, file);
+          if (process.platform === 'win32') {
+            destination = destination.replace(/\\/g, '/');
+          }
+          const uploadOptions = extend({}, {destination}, options);
+          return this.upload(file, uploadOptions).then(
+            () => ({fileName: uploadOptions.destination, status: 'success'}),
```
Comment (callmehiphop) [inline src/bucket.ts]: I feel like we're definitely going to want to pass back the entire response body somehow

File: src/bucket.ts (hunk 12 above)
Comment (stephenplusplus) [reply src/bucket.ts]: Yep, we don't use this anymore.

File: src/bucket.ts (hunk 1 above)
Comment (AVaksman) [reply src/bucket.ts]: Done. Left the types as is till we finalize what this method should return.

File: src/bucket.ts (hunk 2 above)
Comment (AVaksman) [reply src/bucket.ts]: Done, thanks

File: src/bucket.ts (hunk 3 above)
Comment (AVaksman) [reply src/bucket.ts]: Done, thanks

File: src/bucket.ts (hunk 4 above)
Comment (AVaksman) [reply src/bucket.ts]: done

File: src/bucket.ts (hunk 5 above)
Comment (AVaksman) [reply src/bucket.ts]: done

File: src/bucket.ts (hunk 6 above)
Comment (AVaksman) [reply src/bucket.ts]: done

File: src/bucket.ts (hunk 8 above)
Comment (AVaksman) [reply src/bucket.ts]: yep, removed

File: src/bucket.ts (hunk 12 above)
Comment (AVaksman) [reply src/bucket.ts]: removed

File: src/bucket.ts (hunk 11 above)
Comment (AVaksman) [reply src/bucket.ts]: changed

File: src/bucket.ts (hunk 7 above)
Comment (AVaksman) [reply src/bucket.ts]: done

File: src/bucket.ts (hunk 9 above)
Comment (AVaksman) [reply src/bucket.ts]: using new style examples for docs

File: src/bucket.ts (hunk 10 above)
Comment (AVaksman) [reply src/bucket.ts]: Yes, the client library always requests data from the server "as is" via the [headers](https://github.com/googleapis/nodejs-storage/blob/master/src/file.ts#L1238-L1241) ([docs](https://cloud.google.com/storage/docs/transcoding#decompressive_transcoding)) and automatically decompresses if [`isCompressed`](https://github.com/googleapis/nodejs-storage/blob/master/src/file.ts#L1306)

File: src/bucket.ts (hunk 14 above)
Comment (AVaksman) [reply src/bucket.ts]: Do you prefer something like
```ts
return this.upload(file, uploadOptions).then(
            (res) => ({fileName: uploadOptions.destination, status: res[0]}),
            (err: Error) => ({fileName: uploadOptions.destination, status: err})
          );
```
`File` object contains `metadata` property which is the `Metadata` part of the `bucket.upload` response. So essentially we will be returning entire `bucket.upload` response.
Please share you thoughts.

Comment (callmehiphop) [reply src/bucket.ts]: Ah ok, that sounds good to me! Thanks for the break down

File: src/bucket.ts [hunk 15]
```diff
@@ -3174,6 +3189,177 @@ class Bucket extends ServiceObject {
...
+      throw new Error(`${directoryPathString} is an invalid directory`);
+    }
+
+    const fileList: string[] = [];
+    const pathDirName = path.dirname(directoryPathString);
+
+    getFileList(directoryPathString);
+
+    function getFileList(directory: string) {
+      const items = fs.readdirSync(directory);
```
Comment (callmehiphop) [inline src/bucket.ts]: I think we want to use the async versions of all the `fs` methods, I believe the sync versions are all blocking operations.

File: src/bucket.ts [hunk 16]
```diff
@@ -3174,6 +3189,177 @@ class Bucket extends ServiceObject {
...
+        const stat = fs.lstatSync(fullPath);
+        if (stat.isFile()) {
+          fileList.push(fullPath);
+        } else if (options.recurse && stat.isDirectory()) {
+          getFileList(fullPath);
+        }
+      });
+    }
+
+    const limit = pLimit(10);
```
Comment (callmehiphop) [inline src/bucket.ts]: I think we should make this number configurable by the user

**Pull request closed without merging**